### PR TITLE
Login detection setting type migration

### PR DESCRIPTION
**Pull request by nalcalag**
Task/Issue URL: https://app.asana.com/0/72649045549333/1201953975647701/f

### Description
It will make a database upgrade to map `KEY_LOGIN_DETECTION_ENABLED` shared preference from Boolean to enum across the app.
Automatic fireproof experiment is also removed from the code.

### Steps to test this PR

**_Migration in database upgrade_**

_Scenario 1_: User has login detection enabled

- Make sure you have installed a previous version of the app
- Install from this branch
- Open file Device Explorer and go to file: `/data/data/com.duckduckgo.mobile.android.debug/shared_prefs/com.duckduckgo.app.settings_activity.settings.xml`
- Check `KEY_LOGIN_DETECTION_ENABLED` value has been migrated to `ASK_EVERY_TIME`

_Scenario 2_: User has NOT login detection enabled

- Make sure you have installed a previous version of the app
- Install from this branch
- Open file Device Explorer and go to file: `/data/data/com.duckduckgo.mobile.android.debug/shared_prefs/com.duckduckgo.app.settings_activity.settings.xml`
- Check `KEY_LOGIN_DETECTION_ENABLED` value has been migrated to `NEVER`

**_Login detection works with enum_**
- Install from this branch
- Go to Settings > Fireproof sites
- Tap on the toggle button and check it works as expected:
      1. Go to any website where you sign in (e.g. gmail.com)
      2. If toggle is on -> Check current fireproof dialog appears once you sign in
      3. If toggle is off -> Check current fireproof dialog doesn't show when you sign in

### CODE REVIEW

File: app/src/androidTest/java/com/duckduckgo/app/browser/BrowserTabViewModelTest.kt [hunk 1]
```diff
@@ -118,7 +115,12 @@ import com.duckduckgo.app.trackerdetection.model.TrackingEvent
 import com.duckduckgo.app.usage.search.SearchCountDao
 import com.duckduckgo.app.widget.ui.WidgetCapabilities
 import com.duckduckgo.feature.toggles.api.FeatureToggle
-import com.duckduckgo.privacy.config.api.*
+import com.duckduckgo.privacy.config.api.AmpLinkInfo
```
Comment (marcosholgado) [inline app/src/androidTest/java/com/duckduckgo/app/browser/BrowserTabViewModelTest.kt:118]: Have a look at your IDE settings, you have a lot imports that should not be changing and are adding unnecessary changes to the PR.

Comment (nalcalag) [reply app/src/androidTest/java/com/duckduckgo/app/browser/BrowserTabViewModelTest.kt:118]: This is probably because I have "Optimize imports" checked so it removes all unused imports. It shouldn't break anything as it only removes imports that we are not using anymore or it will replace * to the specific class so it is lighter. Did you have problems with this setting before?

Comment (marcosholgado) [reply app/src/androidTest/java/com/duckduckgo/app/browser/BrowserTabViewModelTest.kt:118]: You don't need to mark optimize imports (or anything really) because Spotless already does it for you, same with any other formatting stuff. Did you add the pre-commit hook?

File: app/src/main/java/com/duckduckgo/app/browser/BrowserTabViewModel.kt [hunk 2]
```diff
@@ -46,7 +50,63 @@ import com.duckduckgo.app.bookmarks.model.FavoritesRepository
...
+import com.duckduckgo.app.browser.BrowserTabViewModel.Command.ShowDomainHasPermissionMessage
+import com.duckduckgo.app.browser.BrowserTabViewModel.Command.ShowEditSavedSiteDialog
+import com.duckduckgo.app.browser.BrowserTabViewModel.Command.ShowEmailTooltip
+import com.duckduckgo.app.browser.BrowserTabViewModel.Command.ShowErrorWithAction
+import com.duckduckgo.app.browser.BrowserTabViewModel.Command.ShowFileChooser
+import com.duckduckgo.app.browser.BrowserTabViewModel.Command.ShowFireproofWebSiteConfirmation
+import com.duckduckgo.app.browser.BrowserTabViewModel.Command.ShowFullScreen
+import com.duckduckgo.app.browser.BrowserTabViewModel.Command.ShowKeyboard
+import com.duckduckgo.app.browser.BrowserTabViewModel.Command.ShowSavedSiteAddedConfirmation
+import com.duckduckgo.app.browser.BrowserTabViewModel.Command.ShowWebContent
```
Comment (joshliebe) [inline app/src/main/java/com/duckduckgo/app/browser/BrowserTabViewModel.kt:109]: I agree with @marcosholgado on this one, changing imports like this adds quite a lot of noise to the PR

Comment (aitorvs) [reply app/src/main/java/com/duckduckgo/app/browser/BrowserTabViewModel.kt:109]: +1 to @marcosholgado and @joshliebe points here. I think this PR has a ton of noise because it touches imports. Anything related to code-style should be done through spotless and not through bespoke developer's IDE settings.

If the discussion is whether we like star imports or not, and whether we should change them all, that's something we can discuss internally and modify the spotless rules so that it's also done automatically for us.

It's also worth noting that spotless is breaking CI checks here.

File: app/src/main/java/com/duckduckgo/app/settings/db/SettingsDataStore.kt [hunk 3]
```diff
@@ -91,13 +95,9 @@ class SettingsSharedPreferences @Inject constructor(
...
         set(enabled) = preferences.edit { putBoolean(KEY_AUTOCOMPLETE_ENABLED, enabled) }
 
-    override var appLoginDetection: Boolean
-        get() = if (variantManager.isFireproofExperimentEnabled()) {
-            preferences.getBoolean(KEY_LOGIN_DETECTION_ENABLED, false)
-        } else {
-            preferences.getBoolean(KEY_LOGIN_DETECTION_ENABLED, true)
-        }
-        set(enabled) = preferences.edit { putBoolean(KEY_LOGIN_DETECTION_ENABLED, enabled) }
+    override var appLoginDetection: LoginDetectorSetting
```
Comment (joshliebe) [inline app/src/main/java/com/duckduckgo/app/settings/db/SettingsDataStore.kt:98]: `variantManager` should be removed from the constructor as well, it is no longer used

File: app/src/androidTest/java/com/duckduckgo/app/browser/BrowserTabViewModelTest.kt (hunk 1 above)
Comment (nalcalag) [reply app/src/androidTest/java/com/duckduckgo/app/browser/BrowserTabViewModelTest.kt:118]: Yes, I'm also so used to do control+option+O when I delete things so imports that are not used anymore get removed. I normally also check "Optimize imports" while committing in the IDE. I think I need to check how you use Spotless so I am in the same page as you guys.

File: app/src/androidTest/java/com/duckduckgo/app/fire/fireproofwebsite/ui/FireproofWebsitesViewModelTest.kt [hunk 4]
```diff
@@ -152,6 +153,18 @@ class FireproofWebsitesViewModelTest {
         assertTrue(viewStateCaptor.value.fireproofWebsitesEntities.size == 1)
     }
 
+    @Test
+    fun whenUserTapsOnFireproofSettingsSelectionThenShowFireproofOptionsDialog() {
+
+    }
```
Comment (joshliebe) [inline app/src/androidTest/java/com/duckduckgo/app/fire/fireproofwebsite/ui/FireproofWebsitesViewModelTest.kt:159]: Should this test be implemented?

File: app/src/main/java/com/duckduckgo/app/browser/BrowserTabViewModel.kt (hunk 2 above)
Comment (nalcalag) [reply app/src/main/java/com/duckduckgo/app/browser/BrowserTabViewModel.kt:109]: Sure, I agree that I should delete imports that I'm not using anymore because all the deletions I made, but not modify existing ones through the IDE. I'll change them, thanks! 👍

File: app/src/androidTest/java/com/duckduckgo/app/fire/fireproofwebsite/ui/FireproofWebsitesViewModelTest.kt (hunk 4 above)
Comment (nalcalag) [reply app/src/androidTest/java/com/duckduckgo/app/fire/fireproofwebsite/ui/FireproofWebsitesViewModelTest.kt:159]: No, it is a future test that shouldn't be there because it can't be implemented yet. I will remove it for now

File: app/src/main/java/com/duckduckgo/app/global/db/AppDatabase.kt [hunk 5]
```diff
@@ -552,6 +554,19 @@ class MigrationsProvider(val context: Context) {
...
+    val MIGRATION_42_TO_43: Migration = object : Migration(42, 43) {
+        val fireproofLoginDetector = OldSettingsDataStore()
+
+        override fun migrate(database: SupportSQLiteDatabase) {
+            fireproofLoginDetector.updateFireproofSettingType()
+
+            database.execSQL(
+                "UPDATE $USER_STAGE_TABLE_NAME SET appStage = \"${AppStage.ESTABLISHED}\" " +
+                    "WHERE appStage = \"${AppStage.DAX_ONBOARDING}\""
+            )
```
Comment (joshliebe) [inline app/src/main/java/com/duckduckgo/app/global/db/AppDatabase.kt:566]: I think this was possibly copied by accident from `MIGRATION_41_TO_42` (Or is there a reason we need to skip the onboarding stage?)

File: app/src/main/java/com/duckduckgo/app/global/db/AppDatabase.kt [hunk 6]
```diff
@@ -552,6 +554,19 @@ class MigrationsProvider(val context: Context) {
         }
     }
 
+    val MIGRATION_42_TO_43: Migration = object : Migration(42, 43) {
+        val fireproofLoginDetector = OldSettingsDataStore()
```
Comment (joshliebe) [inline app/src/main/java/com/duckduckgo/app/global/db/AppDatabase.kt:558]: Maybe rename the variable `fireproofLoginDetector` and inner class `OldSettingsDataStore` to something more obvious like: `val fireproofSettingsConverter = FireproofSettingsConverter()`

Comment (nalcalag) [reply app/src/main/java/com/duckduckgo/app/global/db/AppDatabase.kt:558]: Yeah, I agree it will improve readability. Although, I called it `OldSettingsDataStore` to reuse it in case we need to migrate more settings from this class and to be consistency with the other similar class (OldOnboardingStore) in the same file. What do you think? Even if we decide to keep that class name, I would need to change the variable name as it is very confusing, not sure why I called it `fireproofLoginDetector` 🤷‍♀️

Comment (joshliebe) [reply app/src/main/java/com/duckduckgo/app/global/db/AppDatabase.kt:558]: Okay, I didn't see that a naming convention had been set with `OldOnboardingStore` (Not my favorite but I'm all for consistency 😅).

Just change the variable name then 👍

**Comment by joshliebe** [comment]
Looking good so far 👍

**Review by joshliebe — APPROVED** [review]
Tested and works as expected! Good job 👍
